Add HFJetCleaner module for HLT

alpakpinar opened the pull request:
#### PR description:

This PR includes a new forward-jet cleaner module for HLT. Using the reconstructed jets and MET at HLT level, this module produces and saves a collection of cleaned jets which pass the forward-jet cuts. This module will be used in development of several HLT paths, where filtering is done on the missing HT variable computed from the cleaned set of jets. More information about this development and the cleaning cuts applied can be found in [these](https://indico.cern.ch/event/1125473/contributions/4724108/attachments/2391723/4088717/2022-02-15_hlt_exo.pdf) slides. 

The `HLTHFJetCleaner.h` code is included in the `plugins` sub-directory of `HLTrigger/JetMET` module as templated, and the `SealModule.cc` under `HLTrigger/JetMET` is updated with two related module definitions: One for `reco::CaloJet` and the other for `reco::PFJet`.

#### PR validation:

The module is tested locally and the following are verified by comparing collections of original AK4 jets and the filtered jets (with version [8d6ce5a](https://github.com/cms-sw/cmssw/pull/37539/commits/8d6ce5aabc6aee7e79282ed98240595eae1e0474)):

- The module cuts only on the jets that satisfy the `pt`, `eta` and `deltaPhi(jet,MET)` requirements, as specified in the configuration interface of the module.
- The module does not cut on jets if they have a placeholder HF shape variable value (i.e. `-1`).
- The module cuts appropriately on jets that fail the HF variable cuts, as specified in the configuration interface.

#### if this PR is a backport please specify the original PR and why you need to backport that PR

alpakpinar committed:
Add HLTJetHFCleaner module

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29263

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29263/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29263/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29263/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29263/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

missirol commented:
@alpakpinar , please fix the code-checks error reported by the bot . In your local area, you can do
```sh
cmsenv
scram b code-format
scram b code-checks
scram b # check that the updated code does compile
git add -u :/
git commit -m 'apply code checks'
git push my-cmssw hltHFJetCleaner
```

cmsbuild commented:
A new Pull Request was created by  @alpakpinar  for master.

It involves the following packages:

- HLTrigger/Configuration (**hlt**)
- HLTrigger/JetMET (**hlt**)


@cmsbuild, @missirol, @Martin-Grunewald can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @silviodonato this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented on HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
```diff
@@ -0,0 +1,44 @@
+import FWCore.ParameterSet.Config as cms
```
Please remove this file. TSG normally doesn't keep/maintain utilities specific to any one POG/PAG, unless there are special needs (and this does not look like one).

alpakpinar replied:
Implemented in [7547ef9](https://github.com/cms-sw/cmssw/pull/37539/commits/7547ef952a88ab8f3fb13634dabd73d0efc127ea)

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+public:
+  typedef std::vector<reco::PFJet> JetCollection;
+  explicit HLTHFJetCleaner(const edm::ParameterSet&);
+  ~HLTHFJetCleaner() override;
```
```suggestion
  ~HLTHFJetCleaner() override = default;
```

alpakpinar replied:
Implemented in [1cff316](https://github.com/cms-sw/cmssw/pull/37539/commits/1cff316fd36bd151dc2ad6f329f7dfd8e9351b2d)

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+
+class HLTHFJetCleaner : public edm::stream::EDProducer<> {
+public:
+  typedef std::vector<reco::PFJet> JetCollection;
```
Can this plugin make sense for non-PF jets (e.g. Calo) ? If so, it should be templated. If templated, all should go under `plugins/*.h`, and the instantiations in `plugins/SealModule.cc`.

alpakpinar replied:
Thanks for the comments, this plugin is meant to be run on PF jets that are reconstructed during the HLT path sequence, that's why I specified `reco::PFJet` in the code. But the methods accessed for jets are implemented in the base class, `reco::Jet`, so for better generalization, would it work if I replace the `reco::PFJet` objects with the base `reco::Jet` object?

missirol replied:
This should work, or at least, I would try it.

(if there is an expectation that this plugin will evolve in a way that reco::Jet wont suffice anymore, for example if one will need the option to access class members only available for PFJets, then it should be templated; but if this is not expected to be needed for now, reco::Jet would be the simplest option)

alpakpinar replied:
Hi @missirol, while running locally, I've encountered an issue if I use the `reco::Jet` class. I believe the issue is that the module tries to declare a vector of `reco::Jet` objects. Here is the error message:

```
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<std::vector<reco::Jet> >
  std::vector<reco::Jet>
```

This problem does not happen with `reco::PFJet`. Do you know if there is a workaround for this? If not, would it be possible to stick with `reco::PFJet`, given that we intend to run the plugin on those objects. Thanks!

missirol replied:
I think the bottom line (which I partly overlooked in one of my comments yesterday) is that we likely want to produce as output the same type of collection which is used in input (e.g. `reco::PFJet`).

> If not, would it be possible to stick with reco::PFJet, given that we intend to run the plugin on those objects.

Sure, but in that case I would prefer to also support `reco::CaloJet` as usually done for similar plugins. I suggest to do the following.

 - Template the class with 1 template argument (the "jet_type") used to define the type of input and output jets. The whole template will live in one header file, named `plugins/HLTJetHFCleaner.h`.

 - Add 2 instantiations of this template in `plugins/SealModule.cc`, similarly to [this](https://github.com/cms-sw/cmssw/blob/f6c204bfb2f92e827bac326385786fed876046e0/HLTrigger/JetMET/plugins/SealModule.cc#L252), yielding `HLTCaloJetHFCleaner` and `HLTPFJetHFCleaner`.

alpakpinar replied:
I understand, thanks I can do that! So I'll include the templated class under `plugins/HLTJetHFCleaner.h`, and add 2 
2 instantiations under `plugins/SealModule.cc`. I can do something similar to:
```cpp
typedef HLTCaloJetHFCleaner HLTJetHFCleaner<reco::CaloJet>;
DEFINE_FWK_MODULE(HLTCaloJetHFCleaner);

typedef HLTPFJetHFCleaner HLTJetHFCleaner<reco::PFJet>;
DEFINE_FWK_MODULE(HLTPFJetHFCleaner);
```

alpakpinar replied:
Thanks for the suggestion, this is implemented in [5cb552b](https://github.com/cms-sw/cmssw/pull/37539/commits/5cb552b77e666f204527ed68910d6b9a4ead4119). Please let me know if you'd like any changes.

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+  std::vector<T> extractValueMapParameters(edm::Handle<edm::ValueMap<T>> valueMapHandle, size_t numJets);
+
+  edm::EDGetTokenT<edm::View<reco::PFJet>> jetToken_;
+  edm::EDGetTokenT<edm::View<reco::CaloMET>> metToken_;
```
Same comment on (Calo)MET regarding templating. Does this needs to be CaloMET? Or can it be the base class reco::MET ? Or should this be templated to support also other objects like reco::PFMET?

It looks like reco::MET would be sufficient here.

alpakpinar replied:
Similar to the jets case, this should be updated to be the base class `reco::MET` if that works.

alpakpinar replied:
Implemented in [2a4e694](https://github.com/cms-sw/cmssw/pull/37539/commits/2a4e6947671e4c8730dd5dd4c1813d74ee793c1c)

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+      bool passSigmaEtaPhiCut = sieie - sipip < sigmaEtaPhiDiffThresh_;
+
+      if (applyCornerCut_) {
+        bool inCorner = (v_sigmaEtaEta[ijet] < 0.02) & (v_sigmaPhiPhi[ijet] < 0.02);
```
`0.02` is hard-coded. Is it well-motived to hard-code it (e.g. immutable value related to the detector)? If not, please make it configurable via the ParameterSet.

alpakpinar replied:
The `0.02` value was motivated by the analysis we did on these HF shape variables, you can see slide 10 [here](https://indico.cern.ch/event/1125473/contributions/4724108/attachments/2391723/4088717/2022-02-15_hlt_exo.pdf). But I agree that in principle it would be better to make it configurable via the ParameterSet, I'll update that.

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
```
This needs to be updated to include all the configuration parameters of this plugin, each with its default value. Otherwise, (among other things) ConfDB will never know them.

alpakpinar replied:
Implemented in [2a4e694](https://github.com/cms-sw/cmssw/pull/37539/commits/2a4e6947671e4c8730dd5dd4c1813d74ee793c1c)

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
```
```suggestion
  descriptions.addDefaultWithLabel(desc);
```

alpakpinar replied:
Just to clarify, do you mean the method named `addWithDefaultLabel`?

missirol replied:
Yep, apologies.

alpakpinar replied:
Sure no worries, this is implemented in [b130eb6](https://github.com/cms-sw/cmssw/pull/37539/commits/b130eb6e1e6ac2f3287a6a77a5a98b84b37cf069)

missirol reviewed (commented):
A first set of comments.

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+// class declaration
+//
+
+class HLTHFJetCleaner : public edm::stream::EDProducer<> {
```
Make this a `global` module please, rather than a `stream` one. For the syntax, it's just a few changes, and many examples are available in CMSSW, e.g. [this one](https://github.com/cms-sw/cmssw/blob/master/HLTrigger/JetMET/plugins/HLTMultiplicityValueProducer.h).

alpakpinar replied:
Implemented in [719d5ca](https://github.com/cms-sw/cmssw/pull/37539/commits/719d5ca05f4a004dfd43eeacec5ef94d5985f383)

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+  edm::EDGetTokenT<edm::View<reco::PFJet>> jetToken_;
+  edm::EDGetTokenT<edm::View<reco::CaloMET>> metToken_;
+
+  edm::EDGetTokenT<edm::ValueMap<float>> sigmaEtaEtaToken_;
+  edm::EDGetTokenT<edm::ValueMap<float>> sigmaPhiPhiToken_;
+  edm::EDGetTokenT<edm::ValueMap<int>> centralEtaStripSizeToken_;
+
+  // Minimum jet pt to consider while applying HF cuts
+  float jetPtThresh_;
+  float dphiJetMetThresh_;
+
+  // Cuts will be applied to jets only with jetEtaMin_ < |eta| < jetEtaMax_
+  float jetEtaMin_;
+  float jetEtaMax_;
+
+  float sigmaEtaPhiDiffThresh_;
+  float centralEtaStripSizeThresh_;
+  bool applyCornerCut_;
+  bool applySigmaEtaPhiCut_;
+  bool applyStripSizeCut_;
```
Make class data members `const` wherever possible.

alpakpinar replied:
Implemented in [fab4d48](https://github.com/cms-sw/cmssw/pull/37539/commits/fab4d48424a5f8f5544e8a4cc678fc5ce9167998)

missirol commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,268 @@
+// member functions
+//
+
+template <typename T>
+std::vector<T> HLTHFJetCleaner::extractValueMapParameters(edm::Handle<edm::ValueMap<T>> valueMapHandle,
+                                                          size_t numJets) {
+  // Extract the values as a vector from a given ValueMap object.
+  std::vector<T> values;
+  values.reserve(numJets);
+  for (auto it = valueMapHandle->begin(); it != valueMapHandle->end(); it++) {
+    auto productId = it.id();
+    for (size_t i = 0; i < numJets; i++) {
+      T value = valueMapHandle->get(productId, i);
+      values.push_back(value);
+    }
+  }
+  return values;
+}
```
It would be simpler and safer to access the ValueMap via a reference to the object itself, e.g. `vmap[myJetRef]`, also avoiding additional work with indices, and not having to assume that the order of the objects is the same as the order of the VMap entries.

This is a random example: https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoMET/METProducers/src/BeamHaloSummaryProducer.cc#L158

alpakpinar replied:
Thanks for the comment, indeed it would be simpler and safer. Can I ask something though, how would you recommend to obtain `myJetRef`? I tried the following as a first approach but it didn't work:

```cpp
for (size_t ijet = 0; ijet < jets->size(); ijet++) {
    reco::PFJet& jet = (*jets)[ijet]; // Raises compiler error
    float sigmaEtaEta = (*sigmaEtaEtas)[jet];
}
```

I also had some problems with `edm::Ref` (following the example you sent). Is it possible if you can help with that? Thanks so much!

missirol replied:
I haven't tested this, but here is what I would try.

```cpp
// in the class body
using JetCollection = std::vector<jet_type>;
using JetRef = edm::Ref<JetCollection>;

[..]
// in ::produce

auto const h_jets = iEvent.getHandle(jetsToken_); // edm::Handle

for(uint idx=0; idx<h_jets->size(); ++idx){
  JetRef const jetRef(h_jets, idx);
  cout << vmap[jetRef] << endl; // example
}
```

alpakpinar replied:
Thanks, I will try this approach.

alpakpinar replied:
Thanks for the help, this approach worked and I implemented this in [3ca0032](https://github.com/cms-sw/cmssw/pull/37539/commits/3ca0032ab77b75be849b9a3a603a52e6d16ebdf0). 

I just had to change the type of jets handle: `edm::Handle<edm::View>` to `edm::Handle<std::vector>` to make the `edm::Ref` instantiation work. I also removed the now-redundant `extractValueMapParameters` function. Let me know if you want any changes, thanks!

missirol reviewed (commented):
A few technical comments worth giving right away (minor technical comments will come later on).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29269

- This PR adds an extra 24KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
    - Added: 765c323704d14a27da563a99f0e878db915ebac7
    - Modified: bd445c708db87235904aa2827d89dca005fb6ac8
    - Deleted: 7547ef952a88ab8f3fb13634dabd73d0efc127ea
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29271

- This PR adds an extra 24KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
    - Added: 765c323704d14a27da563a99f0e878db915ebac7
    - Modified: bd445c708db87235904aa2827d89dca005fb6ac8
    - Deleted: 7547ef952a88ab8f3fb13634dabd73d0efc127ea
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

Martin-Grunewald commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,269 @@
+//
+/**\class HLTHFJetCleaner HLTHFJetCleaner.cc HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc
+
+ Description: [one line class summary]
```
Please provide a more detailed comment what this plugin does with the input collections etc. and what the decision is based on.

alpakpinar replied:
Implemented in [e4f06dd](https://github.com/cms-sw/cmssw/pull/37539/commits/e4f06dd8ceabde9707fda329e838c4b2bbae193c)

Martin-Grunewald commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,269 @@
+  //virtual void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  //virtual void endRun(edm::Run const&, edm::EventSetup const&) override;
+  //virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+  //virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
```
Remove commented-out lines

alpakpinar replied:
Implemented in [65bcbcc](https://github.com/cms-sw/cmssw/pull/37539/commits/65bcbccb4d1e7fe0a8b6b2ea2d8835d03d8b6e99)

Martin-Grunewald commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,269 @@
+{
+}
+*/
+
```
Remove commented-out blocks.

alpakpinar replied:
Implemented in [65bcbcc](https://github.com/cms-sw/cmssw/pull/37539/commits/65bcbccb4d1e7fe0a8b6b2ea2d8835d03d8b6e99)

Martin-Grunewald commented on HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
```diff
@@ -0,0 +1,269 @@
+  desc.add<edm::InputTag, std::string>("jets", edm::InputTag("hltAK4PFJetsTightIDCorrected"));
+  desc.add<edm::InputTag, std::string>("mets", edm::InputTag("hltMet"));
+  desc.add<edm::InputTag, std::string>("sigmaEtaEta", edm::InputTag("hfJetShowerShapeProducer", "sigmaEtaEta"));
+  desc.add<edm::InputTag, std::string>("sigmaPhiPhi", edm::InputTag("hfJetShowerShapeProducer", "sigmaPhiPhi"));
```
At HLT, producers start with hlt..., not hfJetShower... - adjust the default to work with the intended HLT collection.

alpakpinar replied:
Implemented in [845b43c](https://github.com/cms-sw/cmssw/pull/37539/commits/845b43ca542156c3fb949913cbfc9474949432d8)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29281

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
    - Added: 765c323704d14a27da563a99f0e878db915ebac7
    - Modified: bd445c708db87235904aa2827d89dca005fb6ac8
    - Deleted: 7547ef952a88ab8f3fb13634dabd73d0efc127ea
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29288

- This PR adds an extra 56KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
    - Added: 765c323704d14a27da563a99f0e878db915ebac7
    - Modified: bd445c708db87235904aa2827d89dca005fb6ac8
    - Deleted: 7547ef952a88ab8f3fb13634dabd73d0efc127ea
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
    - Added: b018975d4a10262d29c8fb2037981bc1d8a7326b
    - Modified: 613bc904686ddac8caec8831b5ca7174a2716928, 1cff316fd36bd151dc2ad6f329f7dfd8e9351b2d, 15bc10e560676cd65c28bf4daa1fd60c24d5369d, 4fb7093a63b3e8df1006707f15dd7d55c7238187, 2a4e6947671e4c8730dd5dd4c1813d74ee793c1c, b130eb6e1e6ac2f3287a6a77a5a98b84b37cf069, 719d5ca05f4a004dfd43eeacec5ef94d5985f383, fab4d48424a5f8f5544e8a4cc678fc5ce9167998, 65bcbccb4d1e7fe0a8b6b2ea2d8835d03d8b6e99, e4f06dd8ceabde9707fda329e838c4b2bbae193c, 845b43ca542156c3fb949913cbfc9474949432d8
    - Deleted: 5cb552b77e666f204527ed68910d6b9a4ead4119

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+  produces<JetCollection>();
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
```
```suggestion
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+  descriptions.addWithDefaultLabel(desc);
+}
+
+#endif
```
add newline

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+#ifndef HLTrigger_JetMET_HLTHFJetCleaner_h
```
As written earlier, this belongs in `plugins/` not `interface/`.

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+// ------------ method called to produce the data  ------------
+template <typename JetType>
+void HLTHFJetCleaner<JetType>::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  edm::Handle<std::vector<JetType>> jets;
+  edm::Handle<edm::View<reco::MET>> mets;
+
+  edm::Handle<edm::ValueMap<float>> sigmaEtaEtas;
+  edm::Handle<edm::ValueMap<float>> sigmaPhiPhis;
+  edm::Handle<edm::ValueMap<int>> centralEtaStripSizes;
+
+  iEvent.getByToken(jetToken_, jets);
+
+  iEvent.getByToken(sigmaEtaEtaToken_, sigmaEtaEtas);
+  iEvent.getByToken(sigmaPhiPhiToken_, sigmaPhiPhis);
+  iEvent.getByToken(centralEtaStripSizeToken_, centralEtaStripSizes);
```
When the handle is needed use `iEvent.getHandle(token)`. When the Handle is not needed, use `iEvent.get(token)`: [example](https://github.com/cms-sw/cmssw/blob/3ca0032ab77b75be849b9a3a603a52e6d16ebdf0/HLTrigger/JetMET/plugins/HLTJetTimingFilter.h#L68).

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    JetType jet = (*jets)[ijet];
+    bool withinEtaRange = (abs(jet.eta()) < jetEtaMax_) & (abs(jet.eta()) > jetEtaMin_);
+    // Cuts are applied to jets that are back to back with MET
+    iEvent.getByToken(metToken_, mets);
```
- get MET outside the jet loop..
 - check if MET.size is different than 1 and throw a cms::Exception otherwise (or else, `front()` wont work well)

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+
+  // Loop over the jets and do cleaning
+  for (uint ijet = 0; ijet < jets->size(); ijet++) {
+    JetType jet = (*jets)[ijet];
```
```suggestion
    auto const jet = (*jets)[ijet];
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+  // Loop over the jets and do cleaning
+  for (uint ijet = 0; ijet < jets->size(); ijet++) {
+    JetType jet = (*jets)[ijet];
+    bool withinEtaRange = (abs(jet.eta()) < jetEtaMax_) & (abs(jet.eta()) > jetEtaMin_);
```
```suggestion
    auto const withinEtaRange = (std::abs(jet.eta()) < jetEtaMax_) and (std::abs(jet.eta()) > jetEtaMin_);
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    // Cuts are applied to jets that are back to back with MET
+    iEvent.getByToken(metToken_, mets);
+    const float met_phi = mets->front().phi();
+    bool backToBackWithMet = deltaPhi(jet.phi(), met_phi) > dphiJetMetThresh_;
+    if (!backToBackWithMet) {
```
```suggestion
    if (reco::deltaPhi(jet.phi(), met_phi) <= dphiJetMetThresh_) {
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    const float met_phi = mets->front().phi();
+    bool backToBackWithMet = deltaPhi(jet.phi(), met_phi) > dphiJetMetThresh_;
+    if (!backToBackWithMet) {
+      cleaned_jets->push_back(jet);
```
```suggestion
      cleaned_jets->emplace_back(jet);
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    }
+    // If the jet is outside the range that we're interested in,
+    // append it to the cleaned jets collection and continue
+    if ((jet.pt() < jetPtThresh_) | (!withinEtaRange)) {
```
```suggestion
    if ((jet.pt() < jetPtThresh_) or (!withinEtaRange)) {
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    // If the jet is outside the range that we're interested in,
+    // append it to the cleaned jets collection and continue
+    if ((jet.pt() < jetPtThresh_) | (!withinEtaRange)) {
+      cleaned_jets->push_back(jet);
```
```suggestion
      cleaned_jets->emplace_back(jet);
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    JetRef const jetRef(jets, ijet);
+
+    // Extract HF shape variables from the ValueMap inputs
+    float sigmaEtaEta = (*sigmaEtaEtas)[jetRef];
+    float sigmaPhiPhi = (*sigmaPhiPhis)[jetRef];
+    int centralEtaStripSize = (*centralEtaStripSizes)[jetRef];
```
Use `auto const`. Also, looks like for VMap, handles are not needed (so they can be created via `iEvent.get`).

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+    // We'll apply the sigma eta+phi dependent cut only if requested via the config file
+    if (applySigmaEtaPhiCut_) {
+      // Sigma eta eta and sigma phi phi
+      bool passSigmaEtaPhiCut = sigmaEtaEta - sigmaPhiPhi < sigmaEtaPhiDiffThresh_;
```
```suggestion
      auto const passSigmaEtaPhiCut = (sigmaEtaEta - sigmaPhiPhi) < sigmaEtaPhiDiffThresh_;
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+      bool passSigmaEtaPhiCut = sigmaEtaEta - sigmaPhiPhi < sigmaEtaPhiDiffThresh_;
+
+      if (applyCornerCut_) {
+        bool inCorner = (sigmaEtaEta < cornerCutSigmaEtaEta_) & (sigmaPhiPhi < cornerCutSigmaPhiPhi_);
```
```suggestion
        auto const inCorner = (sigmaEtaEta < cornerCutSigmaEtaEta_) and (sigmaPhiPhi < cornerCutSigmaPhiPhi_);
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+
+      if (applyCornerCut_) {
+        bool inCorner = (sigmaEtaEta < cornerCutSigmaEtaEta_) & (sigmaPhiPhi < cornerCutSigmaPhiPhi_);
+        passSigmaEtaPhiCut = passSigmaEtaPhiCut & (!inCorner);
```
```suggestion
        passSigmaEtaPhiCut &= !inCorner;
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+
+    if (applyStripSizeCut_) {
+      // Cut related to central strip size
+      bool passCentralStripSizeCut = centralEtaStripSize < centralEtaStripSizeThresh_;
+      if (!passCentralStripSizeCut) {
```
```suggestion
      if (centralEtaStripSize >= centralEtaStripSizeThresh_) {
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+template <typename JetType>
+void HLTHFJetCleaner<JetType>::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
```
```suggestion
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+  iEvent.put(std::move(cleaned_jets));
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
```
```suggestion
```

missirol commented on HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
```diff
@@ -0,0 +1,224 @@
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag, std::string>("jets", edm::InputTag("hltAK4PFJetsTightIDCorrected"));
+  desc.add<edm::InputTag, std::string>("mets", edm::InputTag("hltMet"));
+  desc.add<edm::InputTag, std::string>("sigmaEtaEta", edm::InputTag("hltHFJetShowerShape", "sigmaEtaEta"));
+  desc.add<edm::InputTag, std::string>("sigmaPhiPhi", edm::InputTag("hltHFJetShowerShape", "sigmaPhiPhi"));
+  desc.add<edm::InputTag, std::string>("centralEtaStripSize",
+                                       edm::InputTag("hltHFJetShowerShape", "centralEtaStripSize"));
+  desc.add<double, std::string>("jetPtThresh", 100);
+  desc.add<double, std::string>("dphiJetMetThresh", 2.5);
+  desc.add<double, std::string>("jetEtaMin", 2.9);
+  desc.add<double, std::string>("jetEtaMax", 5.0);
+  desc.add<double, std::string>("sigmaEtaPhiDiffThresh", 0.05);
+  desc.add<double, std::string>("cornerCutSigmaEtaEta", 0.02);
+  desc.add<double, std::string>("cornerCutSigmaPhiPhi", 0.02);
+  desc.add<int, std::string>("centralEtaStripSizeThresh", 3);
+  desc.add<bool, std::string>("applyCornerCut", true);
+  desc.add<bool, std::string>("applySigmaEtaPhiCut", true);
+  desc.add<bool, std::string>("applyStripSizeCut", true);
```
I actually don't know what the 2nd template arg does (here `std::string`). I would say it can be removed.

Ideally, you could add comments to explain the purpose of each parameter: [`example`](https://github.com/cms-sw/cmssw/blob/3ca0032ab77b75be849b9a3a603a52e6d16ebdf0/HLTrigger/JetMET/plugins/HLTScoutingPrimaryVertexProducer.cc#L69).

alpakpinar replied:
Thanks! I saw this from the doxygen page [here](https://cmsdoxygen.web.cern.ch/cmsdoxygen/CMSSW_12_4_0_pre2/doc/html/d2/dcb/classedm_1_1ParameterSetDescription.html), but I believe it works without the second template argument. I removed that and commented the parameters.

missirol reviewed (commented):
Some technical comments.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29290

- This PR adds an extra 68KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/JetMET/interface/HLTJetHFCleaner.h:
    - Added: cce119b9719cc0b20871153cc9ad267580de9a05
    - Modified: c33b53a82c18785bca7b316f846d6da5832cb3bc, 9cf49cc7aef5529cea9e77155c914a9b4904582c, c0299a3fd3b6e9717181e00762f54db24a1f4b46
    - Deleted: 4acd6b1337056dbfd7cb4e8fa727d089afe93e8e
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
    - Added: 765c323704d14a27da563a99f0e878db915ebac7
    - Modified: bd445c708db87235904aa2827d89dca005fb6ac8
    - Deleted: 7547ef952a88ab8f3fb13634dabd73d0efc127ea
  - HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
    - Added: 5cb552b77e666f204527ed68910d6b9a4ead4119
    - Modified: 3ca0032ab77b75be849b9a3a603a52e6d16ebdf0
    - Deleted: cce119b9719cc0b20871153cc9ad267580de9a05
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
    - Added: b018975d4a10262d29c8fb2037981bc1d8a7326b
    - Modified: 613bc904686ddac8caec8831b5ca7174a2716928, 1cff316fd36bd151dc2ad6f329f7dfd8e9351b2d, 15bc10e560676cd65c28bf4daa1fd60c24d5369d, 4fb7093a63b3e8df1006707f15dd7d55c7238187, 2a4e6947671e4c8730dd5dd4c1813d74ee793c1c, b130eb6e1e6ac2f3287a6a77a5a98b84b37cf069, 719d5ca05f4a004dfd43eeacec5ef94d5985f383, fab4d48424a5f8f5544e8a4cc678fc5ce9167998, 65bcbccb4d1e7fe0a8b6b2ea2d8835d03d8b6e99, e4f06dd8ceabde9707fda329e838c4b2bbae193c, 845b43ca542156c3fb949913cbfc9474949432d8
    - Deleted: 5cb552b77e666f204527ed68910d6b9a4ead4119

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,219 @@
+          "corner value.");
+  desc.add<int>("centralEtaStripSizeThresh", 3)
+      ->setComment("Determines the threshold in the following cut: centralEtaStripSize < X");
+  desc.add<bool>("applyCornerCut", true)->setComment("Boolean specifying whether to apply the corner cut.");
```
Does it make sense to name this `applySigmaEtaPhiCornerCut` (or similar)? I'm nitpicking, but `applyCornerCut` only applies if `applySigmaEtaPhiCut_` is true, and it'd be good to sort-of see that from its name.

alpakpinar replied:
Thanks for the comments, I think it makes sense to do that yes, I renamed the parameter in [34a2e2a](https://github.com/cms-sw/cmssw/pull/37539/commits/34a2e2a191571dc91b191212fab00330e984cef7).

missirol commented on HLTrigger/JetMET/plugins/SealModule.cc:
```diff
@@ -256,3 +258,9 @@ DEFINE_FWK_MODULE(HLTPFJetsMatchedToFilteredCaloJetsProducer);
+DEFINE_FWK_MODULE(HLTCaloJetHFCleaner);
+
+typedef HLTJetHFCleaner<reco::PFJet> HLTPFJetHFCleaner;
+DEFINE_FWK_MODULE(HLTPFJetHFCleaner);
```
add newline

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,219 @@
+      }
+    }
+
+    if (applyStripSizeCut_) {
+      // Cut related to central strip size
+      if (centralEtaStripSize >= centralEtaStripSizeThresh_) {
+        continue;
+      }
+    }
```
```suggestion
    // Cut related to central strip size
    if (applyStripSizeCut_ and centralEtaStripSize >= centralEtaStripSizeThresh_) {
        continue;
    }
```

missirol reviewed (commented):
Few more technical comments.

alpakpinar commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,219 @@
+
+  // Check if length of the MET vector is 1, throw exception otherwise
+  if (mets->size() != 1) {
+    throw cms::Exception("ComparisonFailure", "Length of MET vector is different than 1");
```
I think it should be equivalent, but to make sure I replaced this line with your suggestion in [cdb13e6](https://github.com/cms-sw/cmssw/pull/37539/commits/cdb13e6a20657a8cd1e659278f63d91b6f3182cd).

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,219 @@
+    JetRef const jetRef(jets, ijet);
+
+    // Extract HF shape variables from the ValueMap inputs
+    auto const sigmaEtaEta = sigmaEtaEtas[jetRef];
+    auto const sigmaPhiPhi = sigmaPhiPhis[jetRef];
+    auto const centralEtaStripSize = centralEtaStripSizes[jetRef];
```
By the way, these VMap calls are not always needed (as their use depends on flags like `applySigmaEtaPhiCut_`). We can move these line inside the relevant `if` blocks. Even the `consume` calls of the VMaps could be made conditional (meaning, for example, not calling consumes for VMaps that won't be used in ::produce, based on the value of flags).

alpakpinar replied:
Ah sure I can update that. So within the constructor, I can only call the `consume` calls depending on values of corresponding boolean flags. And then in `produce` method, retrieve them only if the flags are `true`.

alpakpinar replied:
I tried to address this in [a1fc8c9](https://github.com/cms-sw/cmssw/pull/37539/commits/a1fc8c97ce8e96cda540d78f8f7d138bb58aa9b3). I moved the `get` methods in corresponding if blocks, but I didn't touch the constructor because of compiler complications about which members getting initialized first (if one of the flags is `False` and the other one is `True`, the order of initializations gets swapped). Let me know what you think and if you want more changes to this, thanks!

missirol replied:
@alpakpinar , thanks for trying. I realise this is not entirely trivial, and I don't want to delay this PR for this (i might decide to address this technical point in the future). [by the way, we would not move the `.get` calls inside the loop]

So, I retract the comment, and suggest to undo a1fc8c97ce8e96cda540d78f8f7d138bb58aa9b3. I left one more comment below, which should be my final one.

One more thing: in your next update, please squash all commits into 1, and force-push to the same branch. The result will be a PR with just 1 commit (the discussion remains on GitHub, and on git we dont need/want to keep centrally all these intermediate commits).

alpakpinar replied:
Ah you're right sorry, I shouldn't have put the `get` call inside the `for` loop. Okay then, thanks so much and I'll add a `revert` commit for this one. After I've addressed your other comments, I'll squash all commits into 1 and force-push as you said.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29310

- This PR adds an extra 56KB to repository

- Found files with invalid states:
  - HLTrigger/HLTHFJetCleaner/test/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/JetMET/interface/HLTJetHFCleaner.h:
    - Added: cce119b9719cc0b20871153cc9ad267580de9a05
    - Modified: c33b53a82c18785bca7b316f846d6da5832cb3bc, 9cf49cc7aef5529cea9e77155c914a9b4904582c, c0299a3fd3b6e9717181e00762f54db24a1f4b46
    - Deleted: 4acd6b1337056dbfd7cb4e8fa727d089afe93e8e
  - HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: 765c323704d14a27da563a99f0e878db915ebac7, d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/Configuration/python/customizeHLTforJetCleaning.py:
    - Added: 765c323704d14a27da563a99f0e878db915ebac7
    - Modified: bd445c708db87235904aa2827d89dca005fb6ac8
    - Deleted: 7547ef952a88ab8f3fb13634dabd73d0efc127ea
  - HLTrigger/JetMET/interface/HLTHFJetCleaner.h:
    - Added: 5cb552b77e666f204527ed68910d6b9a4ead4119
    - Modified: 3ca0032ab77b75be849b9a3a603a52e6d16ebdf0
    - Deleted: cce119b9719cc0b20871153cc9ad267580de9a05
  - HLTrigger/HLTHFJetCleaner/plugins/BuildFile.xml:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Modified: d3155c44d350696b4e60543f5b6fdb174d1f2509
    - Deleted: b018975d4a10262d29c8fb2037981bc1d8a7326b
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_main.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/HLTHFJetCleaner/test/test_catch2_HLTHFJetCleaner.cc:
    - Added: 717eef6bd0c739a3a9389f4b1dfe7fabb251a138
    - Deleted: 03d20824e0a5f93d4faa68a9a0b4ca1e5822590c
  - HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc:
    - Added: b018975d4a10262d29c8fb2037981bc1d8a7326b
    - Modified: 613bc904686ddac8caec8831b5ca7174a2716928, 1cff316fd36bd151dc2ad6f329f7dfd8e9351b2d, 15bc10e560676cd65c28bf4daa1fd60c24d5369d, 4fb7093a63b3e8df1006707f15dd7d55c7238187, 2a4e6947671e4c8730dd5dd4c1813d74ee793c1c, b130eb6e1e6ac2f3287a6a77a5a98b84b37cf069, 719d5ca05f4a004dfd43eeacec5ef94d5985f383, fab4d48424a5f8f5544e8a4cc678fc5ce9167998, 65bcbccb4d1e7fe0a8b6b2ea2d8835d03d8b6e99, e4f06dd8ceabde9707fda329e838c4b2bbae193c, 845b43ca542156c3fb949913cbfc9474949432d8
    - Deleted: 5cb552b77e666f204527ed68910d6b9a4ead4119

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,216 @@
+      ->setComment(
+          "Corner cut value for sigmaPhiPhi. Jets will be cut if both sigmaEtaEta and sigmaPhiPhi are lower than the "
+          "corner value.");
+  desc.add<int>("centralEtaStripSizeThresh", 3)
```
Please change `Thresh` to `Min` or `Max` depending on the cut, to make more obvious in which direction the cut goes..

alpakpinar replied:
Thanks, I updated the parameter names as such. Just to note, when I used `centralEtaStripSizeMax` as a variable, I set the default to `2` instead, and allowed jets to pass `centralEtaStripSize <= centralEtaStripSizeMax`, since I thought `Max` value should be allowed. I did the same implementation for `sigmaEtaPhiDiffMax`. Let me know if you have any comments about this, thanks!

alpakpinar commented:
Hi @missirol, sorry I messed up the first force push (there was an extra commit), but the second force push should have only one commit with all the changes.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29319

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,213 @@
+
+    JetRef const jetRef(jets, ijet);
+
+    auto const& sigmaEtaEtas = iEvent.get(sigmaEtaEtaToken_);
+    auto const& sigmaPhiPhis = iEvent.get(sigmaPhiPhiToken_);
+    auto const& centralEtaStripSizes = iEvent.get(centralEtaStripSizeToken_);
```
Outside the jet loop..

alpakpinar replied:
Sorry about that, also should be fixed in [ff94a85](https://github.com/cms-sw/cmssw/pull/37539/commits/ff94a85983bfb7cbbb37f7dc76656d354bff00d8).

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,213 @@
+    auto const jet = (*jets)[ijet];
+    auto const withinEtaRange = (std::abs(jet.eta()) < jetEtaMax_) and (std::abs(jet.eta()) > jetEtaMin_);
+    // Cuts are applied to jets that are back to back with MET
+    if (reco::deltaPhi(jet.phi(), met_phi) <= dphiJetMetMin_) {
```
Should this be `abs(deltaPhi)` ?

alpakpinar replied:
Thanks a lot for catching this, it should indeed be the `abs` value. I fixed this in [ff94a85](https://github.com/cms-sw/cmssw/pull/37539/commits/ff94a85983bfb7cbbb37f7dc76656d354bff00d8)

missirol reviewed (commented):
Tentatively final comments.

missirol commented:
@alpakpinar , note that the PR tests will not exercise the new plugin in any way. My suggestion is that, once the implementation is final (it almost is), you check carefully that the latest v of the plugin produces the intended results in your HLT tests (once you can give feedback on that, I will sign off on the PR).

alpakpinar commented:
Sounds good @missirol, thanks so much for reviewing the code! I'll check the implementation on some HLT tests I run and I'll let you know, meanwhile if you have more  comments I can work on them too. Also let me know if you want me to squash the second fixup commit and force push again. Thanks!

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29321

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
> I'll check the implementation on some HLT tests I run and I'll let you know, meanwhile if you have more comments I can work on them too. Also let me know if you want me to squash the second fixup commit and force push again. 

Thanks, @alpakpinar . For the squash, I leave it to you (2 commits is fine, it can stay like this).

Please update PR title and description in accordance with the latest version of these updates (name and location of the file changed).

The next CMSSW release is expected to be finalised next Tue, so it would be nice to have feedback by Monday.

alpakpinar commented:
> > I'll check the implementation on some HLT tests I run and I'll let you know, meanwhile if you have more comments I can work on them too. Also let me know if you want me to squash the second fixup commit and force push again.
> 
> Thanks, @alpakpinar . For the squash, I leave it to you (2 commits is fine, it can stay like this).
> 
> Please update PR title and description in accordance with the latest version of these updates (name and location of the file changed).
> 
> The next CMSSW release is expected to be finalised next Tue, so it would be nice to have feedback by Monday.

Thanks @missirol, sounds good, if that's fine then I will leave it in 2 commits. I will do tests with this version of the code, and update the PR description with the new files + the testing I've done. I will try to get this done by tomorrow, but in the worst case I should have this ready by Monday. Thanks again for the review!

Martin-Grunewald commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,213 @@
+
+  // Loop over the jets and do cleaning
+  for (uint ijet = 0; ijet < jets->size(); ijet++) {
+    auto const jet = (*jets)[ijet];
```
Replace two lines by:
```for (auto const & jet : *jets) {```

Martin-Grunewald replied:
Ah sorry, you need the ```ijet``` counter below... never mind.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29345

- This PR adds an extra 16KB to repository

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,221 @@
+
+      // Check if sigma eta eta and sigma phi phi are both set to -1.
+      // If this is the case, we can skip the jet without evaluating any mask.
+      if ((sigmaEtaEta == -1) and (sigmaPhiPhi == -1)) {
```
I don't know if this can happen here, but in principle this comparison between floats could fail if the valueMap returns something like `-1.0000002` (for floating-point precision, or whatever other reason). Would it make sense to ask for `< 0` instead?

alpakpinar replied:
Ah makes much more sense, sure I'll do that, thanks!

alpakpinar replied:
I did this, squashed that last commit to the previous one, and pushed the merged commit, [1535d04](https://github.com/cms-sw/cmssw/pull/37539/commits/1535d04aa40544ecdf0579075ff871e61f4b718f). Let me know if you have more comments, thanks!

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

alpakpinar commented:
Hi @missirol, while examining the code, I found an issue on the case where the jet has a `-1` value for `sigmaEtaEta` and `sigmaPhiPhi`. In that case, the `inCorner` boolean flag results to `true`, which results in the jet being removed from the collection, even though it was not meant to. I fixed that problem in  [122331d](https://github.com/cms-sw/cmssw/pull/37539/commits/122331da305e5dbc596d85112a9ad73eb851c6e2), and updated the PR description with the tests. 

With this latest version, everything looks to be working as intended. Let me know if you have any questions or comments, thanks!

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29347

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on HLTrigger/JetMET/plugins/HLTJetHFCleaner.h:
```diff
@@ -0,0 +1,221 @@
+
+      // Check if sigma eta eta and sigma phi phi are both set to -1.
+      // If this is the case, we can skip the jet without evaluating any mask.
+      if ((sigmaEtaEta < 0) and (sigmaPhiPhi < 0)) {
```
Just for completeness, could you elaborate a bit on what can cause these vmaps to return negative values?

Should this plugin also handle the case where only 1 of the 2 is negative? (e.g. `(sigmaEtaEta < 0) or (sigmaPhiPhi < 0)`)

alpakpinar replied:
Sure, so the vmaps are coming from [this](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc) module. And for `sigmaEtaEta` and `sigmaPhiPhi` values, the placeholder value is `-1`. I believe this happens in two cases:

1. The jet is not within the eta range (default is `|eta|>2.9`)
2. The jet is within the eta range, but it has `pt < ptMin`, `ptMin` specified from the configuration of that module
3. These values might still end up as `-1` due to [here](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc#L181-L187). I believe this is the case where the PU corrected candidate `pt`s are less than a threshold specified from the configuration.

So we shouldn't have to handle the case where only one of them is `-1`, because either both values end up as `-1`, or both result in non-dummy values. Let me know if you have further questions or comments, thanks!

missirol replied:
Thanks for the clarification. My only suggestion then would be to add a comment somewhere in this plugin saying that the valuemaps are normally created by the producer [HFJetShowerShape](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc), and take this last tweak as a chance to squash in 1 commit.

alpakpinar replied:
Sure, I've added a comment and squashed everything into one commit in [8d6ce5a](https://github.com/cms-sw/cmssw/pull/37539/commits/8d6ce5aabc6aee7e79282ed98240595eae1e0474).

missirol replied:
One more question on this.

If `applyStripSizeCut_==true` and the vmap values are negative, the jet is retained in L160 regardless of whether or not it passes the strip-size cut. Is this the intended behaviour? (if the jet fails the strip-size cut, shouldn't it be discarded even if its sigma-eta/phi values are negative?).

Going back to the point on `! ((sigmaEtaEta < 0) and (sigmaPhiPhi < 0))` vs `! ((sigmaEtaEta < 0) or (sigmaPhiPhi < 0))`, I think we should go with the latter, because that (i.e. both positive) is the regime for which the cut "sigma_eta - sigma_phi > X" is meaningful. Effectively it makes no difference now, but this would make this plugin more self-consistent.

In general, we now have in the jet loop several repetitions of the pattern (in pseudo-code)
```cpp
 if (dont_cut_on_this_jet or cut_passes) {
      cleaned_jets->emplace_back(jet);
      continue;
 }
```
It might be clearer to organise this in a different way, e.g.
```cpp
for ( jet_loop ) {
  bool keepJet = true;
  if ( pT > A and B<|eta|<C) {
    if ( abs(deltaPhi) > D) {
      if (applySigmaEtaPhiCut_) {
        if (sigmaEtaEta >= 0 and sigmaPhiPhi >= 0) {
          keepJet &= (sigmaEtaEta - sigmaPhiPhi) <= sigmaEtaPhiDiffMax_;
          if (applySigmaEtaPhiCornerCut_) {
            keepJet &= (sigmaEtaEta > cornerCutSigmaEtaEta_) or (sigmaPhiPhi > cornerCutSigmaPhiPhi_);
          }
        }
      }
      if (keepJet and applyStripSizeCut_) {
        keepJet &= (centralEtaStripSizes[jetRef] <= centralEtaStripSizeMax_);
      }
    }
  }
  if (keepJet) {
    cleaned_jets->emplace_back(jet);
  }
}
```
(Note: this is pseudo-code, and I didn't validate any of this; there might be mistakes)

That said, this point could be addressed in a follow-up PR. This PR could go as is in time for the next pre-release (Tue Apr-19) such that the new plugin becomes available in ConfDB.

alpakpinar replied:
You're right, but if the vMap values for `sigmaEtaEta` and `sigmaPhiPhi` are negative, I think it is not possible for a jet to fail the `centralEtaStripSize` cut. Referring to the three cases above, if case 1 or 2 is True, we would have the strip size equal to dummy value, `0`. Case 3 is less obvious, but if all PF candidate `pt`s (PU subtracted) are less than the `widthPtThreshold`, they will also be less than the `stripPtThreshold`, since (at least by default values)  `stripPtThreshold > widthPtThreshold`. (The `stripPtThreshold` cutoff is used for counting the strips and adding them to the strip size variable [here](https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/plugins/HFJetShowerShape.cc#L162-L164).). So the `centralEtaStripSize` for this case would be `0` as well.

Regarding your other comment, I agree that going with `! ((sigmaEtaEta < 0) or (sigmaPhiPhi < 0))` would be more self-consistent, even though it effectively makes no difference now. I also agree that re-organizing the code as you suggested can make things cleaner. But as you pointed out, I can do this re-organization (together with the change in the `if` statement mentioned in the previous sentence) in a follow-up PR if that works. Thanks so much for your review and your comments!

alpakpinar replied:
Regarding ConfDB and trigger studies, I had an additional question. According to [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideGlobalHLT#CMSSW_12_4_X), the trigger studies are done by pulling the latest `CMSSW_12_3_0/GRun` tables, if I understand correctly. For this plugin to be available on those tables, would I need an additional backport PR to `CMSSW_12_3_X`? Thanks so much for the help!

missirol replied:
Sure. That works.

missirol replied:
> Regarding ConfDB and trigger studies, I had an additional question. According to [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideGlobalHLT#CMSSW_12_4_X), the trigger studies are done by pulling the latest `CMSSW_12_3_0/GRun` tables, if I understand correctly. For this plugin to be available on those tables, would I need an additional backport PR to `CMSSW_12_3_X`? Thanks so much for the help!

The answer to the question is 'yes'. On the other hand, I assume these developments are related to [CMSHLT-2282](https://its.cern.ch/jira/browse/CMSHLT-2282), and the latter targets the so-called HLT-V2 menu, which will be in `12_4_X`, so in that sense a backport of this PR to `12_3_X` is not really needed (plus, I would like to avoid 2 backports b/c of the technical tweaks discussed above). The current plan is to move HLT-menus to `12_4_X` after May-2. The normal course of action would be that (1) we integrate this PR now, (2) this producer becomes available in ConfDB after `12_4_0_pre3` is released and you can then use it in your ConfDB areas, and (3) the integration of CMSHLT-2282 will happen once the GRun menu moves to `12_4_X` in early May.

Does this work for you?

alpakpinar replied:
Yes that's right, these developments are related to the HF-filtered MHTNoMu triggers in [CMSHLT-2282](https://its.cern.ch/jira/browse/CMSHLT-2282), and targets the HLT-V2 menu. I understand, thanks for explaining, I think that should work for me. 

So if I understand correctly, the `12_4_0_pre3` release will be available on Apr-19, so that this module becomes available on ConfDB. I can use it, together with the latest `12_3_0/GRun` tables to finalize my studies on rates, timing and integration tests. And we can integrate these paths on early May when the GRun menu moves to `12_4_X`, does that sound good? Thanks so much again for the help!

missirol replied:
> So if I understand correctly, the 12_4_0_pre3 release will be available on Apr-19, so that this module becomes available on ConfDB.

Yes.

> I can use it, together with the latest 12_3_0/GRun tables to finalize my studies on rates, timing and integration tests.

In general, yes. This might not be entirely smooth depending on what exactly you want to do. As a last resort for your studies between now and the end of April, you can always add these producers to a local `12_3_X` area of yours (a local manual backport of sorts). In any case, I'm sure one can figure out a proper way to proceed with studies despite these minor complications; you can always ask in the hlt-support channel in case of doubts.

By the way, before the integration in the central tables, the update of these Paths should be discussed/presented in TSG. Or has that already happened?

alpakpinar replied:
I understand, thanks. Sure then, I can try to do include these producers on my local `12_3_X` area and do the studies as such. I'll contact the Mattermost channel in case of issues, thanks for the help.

I haven't presented this in TSG yet, but I have a talk scheduled on the upcoming meeting, April 20, where I can discuss the cleaning cuts that are applied in this producer, and the checks we did on the impact on noise-enriched events. With this module getting into ConfDB on April 19, I'm not sure if I can get all the studies completed by the next day (rates, timing, integration tests), but would it work if I show some preliminary results I got on Apr 20, and then complete the studies before the end of April? Or would you prefer to postpone the talk until I got those all done, I would be okay with the option that works best for the TSG.

missirol replied:
Understood. I don't do the TSG schedule :) ; in case of doubts, feel free to ask Darin and Silvio by email.

In any case, the producers don't really need to be in ConfDB to be able to extract preliminary numbers for rates, timing, etc. I think you already have everything right now to put together those numbers (one can do everything in `python` manipulating the existing HLT menus). The details needed for the integration (ConfDB area, etc) can certainly be ironed out after a 1st presentation in TSG.

alpakpinar replied:
Ah okay got it, thanks! Sure then I can pull existing HLT menus via `hltGetConfiguration`, and do the updates in the python config file to put together the results for Apr 20. And we can iron out the details about the integration after a 1st presentation as you said, thanks for the help!

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37539/29348

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #37539 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
please test

Thanks, @alpakpinar . I noticed that the comment referring to "122331d" in the PR description should be fixed.

alpakpinar commented:
Thanks for pointing that out @missirol, I updated that to the correct commit hash now.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df751b/23962/summary.html
**COMMIT**: 8d6ce5aabc6aee7e79282ed98240595eae1e0474
**CMSSW**: CMSSW_12_4_X_2022-04-15-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37539/23962/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3593159
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3593123
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-04-15-1100+df751b/49647/triggerResults): no differences found

missirol commented:
+hlt

  - adds a two new HLT producers, not used yet in any central wfs nor any HLT menus
  - technical improvements promised by @alpakpinar [in a follow-up PR](https://github.com/cms-sw/cmssw/pull/37539#discussion_r851781657) (postponed to have the new producer in ConfDB starting from `CMSSW_12_4_0_pre3`)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
